transport/whisper/envelopes.go

```diff
@@ -0,0 +1,324 @@
+			log.Warn("envelope that was posted by us is discarded", "hash", envelopeError.Hash, "peer", event.Peer, "error", envelopeError.Description)
+			var err error
+			switch envelopeError.Code {
+			case whisper.EnvelopeTimeNotSynced:
+				err = errors.New("envelope wasn't delivered due to time sync issues")
+			}
+			m.handleEnvelopeFailure(envelopeError.Hash, err)
+		}
+		failedEnvelopes[envelopeError.Hash] = struct{}{}
+	}
+
+	for hash := range envelopes {
+		if _, exist := failedEnvelopes[hash]; exist {
+			continue
+		}
+		state, ok := m.envelopes[hash]
+		if !ok || state == EnvelopeSent {
+			continue
+		}
+		m.envelopes[hash] = EnvelopeSent
+		if m.handler != nil {
+			m.handler.EnvelopeSent(m.identifiers[hash], messageID(m.messages[hash]))
+		}
+	}
+	delete(m.batches, event.Batch)
+}
+
+func (m *EnvelopesMonitor) handleEventEnvelopeExpired(event whisper.EnvelopeEvent) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.handleEnvelopeFailure(event.Hash, errors.New("envelope expired due to connectivity issues"))
+}
+
+// handleEnvelopeFailure is a common code path for processing envelopes failures. not thread safe, lock
+// must be used on a higher level.
+func (m *EnvelopesMonitor) handleEnvelopeFailure(hash common.Hash, err error) {
+	if state, ok := m.envelopes[hash]; ok {
+		message, exist := m.messages[hash]
+		if !exist {
+			log.Error("message was deleted erroneously", "envelope hash", hash)
+		}
+		mID := messageID(message)
+		attempt := m.attempts[hash]
+		m.clearMessageState(hash)
+		if state == EnvelopeSent {
+			return
+		}
+		if attempt < m.maxAttempts {
+			log.Debug("retrying to send a message", "message id", mID, "attempt", attempt+1)
+			hex, err := m.whisperAPI.Post(context.TODO(), message)
+			if err != nil {
+				log.Error("failed to retry sending message", "message id", mID, "attempt", attempt+1)
+			}
+			envelopeID := common.BytesToHash(hex)
+			m.messageToEnvelope[mID] = envelopeID
+			m.envelopes[envelopeID] = EnvelopePosted
+			m.messages[envelopeID] = message
+			m.attempts[envelopeID] = attempt + 1
+		} else {
+			log.Debug("envelope expired", "hash", hash, "state", state)
+			if m.handler != nil {
+				m.handler.EnvelopeExpired(mID, err)
+			}
+		}
+	}
+}
+
+func (m *EnvelopesMonitor) handleEventEnvelopeReceived(event whisper.EnvelopeEvent) {
+	if m.mailServerConfirmation {
+		if !m.isMailserver(event.Peer) {
+			return
+		}
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	state, ok := m.envelopes[event.Hash]
+	if !ok || state != EnvelopePosted {
+		return
+	}
+	log.Debug("expected envelope received", "hash", event.Hash, "peer", event.Peer)
+	m.envelopes[event.Hash] = EnvelopeSent
+	if m.handler != nil {
+		m.handler.EnvelopeSent(m.identifiers[event.Hash], messageID(m.messages[event.Hash]))
+	}
+}
+
+// clearMessageState removes all message and envelope state.
+// not thread-safe, should be protected on a higher level.
+func (m *EnvelopesMonitor) clearMessageState(envelopeID common.Hash) {
+	delete(m.envelopes, envelopeID)
```

Comment:
Tricky mutex management ;) We can fix that later but generally there is no pattern which method assumes a mutex is already acquired when called and which does it by itself.

Reply:
this is all copied from status-go, haven't looked at the code much to be honest, probably best to fix later as I would have to familiarize myself with the logic